### PR TITLE
[Review Needed]Pytest Style analysis/test_rms.py

### PR DESCRIPTION
**Pull request by utkbansal**
@kain88-de @richardjgowers @jbarnoud Need help with `test_with_superposition_smaller` and `test_with_superposition_equal`

Error for `test_with_superposition_equal`

```
MDAnalysisTests/analysis/test_rms.py:156 (Testrmsd.test_with_superposition_equal)
self = <testsuite.MDAnalysisTests.analysis.test_rms.Testrmsd object at 0x10da1ca90>
p_first = <AtomGroup with 3341 atoms>, p_last = <AtomGroup with 3341 atoms>

    def test_with_superposition_equal(self, p_first, p_last):
        align.alignto(p_first, p_last)
        A = p_first.positions
        B = p_last.positions
        rmsd = rms.rmsd(A, B)
        rmsd_superposition = rms.rmsd(A, B, center=True, superposition=True)
>       assert_almost_equal(rmsd, rmsd_superposition)
E       AssertionError: 
E       Arrays are not almost equal to 7 decimals
E        ACTUAL: 3.3291879348562332e-08
E        DESIRED: 1.2660351092866358e-06

test_rms.py:163: AssertionError 
```

### CODE REVIEW

**Comment by jbarnoud** [comment]
I did not read the all thing, but you have two fixtures called `u2` in the same class (just following each other), it may cause issues.

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 1]
```diff
@@ -47,100 +44,120 @@
...
+
+    @pytest.fixture()
+    def b(self, a):
+        return a + self.ones
+
+    @pytest.fixture()
+    def u(self):
+        u = mda.Universe(PSF, DCD)
+        u.trajectory[2]
+        u.trajectory[0]
```
Comment (kain88-de) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: This has no affect. The trajectory will be at frame 0 no matter what.

File: testsuite/MDAnalysisTests/analysis/test_rms.py [hunk 2]
```diff
@@ -47,100 +44,120 @@
...
+        return u
+
+    @pytest.fixture()
+    def u2(self):
+        u = mda.Universe(PSF, DCD)
+        u.trajectory[-2]
+        u.trajectory[-1]
```
Comment (kain88-de) [inline testsuite/MDAnalysisTests/analysis/test_rms.py]: This will just set you to the last frame. I think you can remove this fixture.

**Comment by kain88-de** [comment]
build fails

**Comment by orbeckst** [comment]
To fix the failures, use decimals=6. It seems that it worked to machine precision before but I have no problem specifying it explicitly or dialing down the precision. Mismatches at the level of 1e-06 do not worry me for the RMSD.

**Comment by orbeckst** [comment]
FYI: If PR #1581 is merged before this one, then you will need to rebase. Otherwise, #1581 will need a rebase (and adjusting to to new pytest style).

**Comment by utkbansal** [comment]
@kain88-de @orbeckst One test is failing
```python
    def test_with_superposition_smaller(self, p_first, p_last):
        A = p_first.positions
        B = p_last.positions
        rmsd = rms.rmsd(A, B)
        rmsd_superposition = rms.rmsd(A, B, center=True, superposition=True)
        print(rmsd, rmsd_superposition)
        # by design the super positioned rmsd is smaller
        assert rmsd > rmsd_superposition
```

pytest error output - 
```
    def test_with_superposition_smaller(self, p_first, p_last):
        A = p_first.positions
        B = p_last.positions
        rmsd = rms.rmsd(A, B)
        rmsd_superposition = rms.rmsd(A, B, center=True, superposition=True)
        print(rmsd, rmsd_superposition)
        # by design the super positioned rmsd is smaller
>       assert rmsd > rmsd_superposition
E       assert 0.0 > 1.0884445076843497e-06
```

**Comment by orbeckst** [comment]
@utkbansal ,  PR #1581 was merged. Did you already rebase this branch against develop? If not, please do this first. 

It only makes sense to debug after this rebase.

**Comment by utkbansal** [comment]
Yes, I've already done the rebase.

On 20-Aug-2017 11:47 AM, "Oliver Beckstein" <notifications@github.com>
wrote:

@utkbansal <https://github.com/utkbansal> , PR #1581
<https://github.com/MDAnalysis/mdanalysis/pull/1581> was merged. Did you
already rebase this branch against develop? If not, please do this first.

It only makes sense to debug after this rebase.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/MDAnalysis/mdanalysis/pull/1610#issuecomment-323566289>,
or mute the thread

**Comment by orbeckst** [comment]
Ok, will try to have a look today. I'm traveling so will depend a little bit on how the traveling goes. 

> Am Aug 20, 2017 um 00:09 schrieb Utkarsh Bansal <notifications@github.com>:
> 
> Yes, I've already done the rebase.

**Comment by orbeckst** [comment]
I can reproduce the failure locally (Python 3.6, macOS 10.12.6)
```
>       assert rmsd > rmsd_superposition
E       assert 0.0 > 1.0884445076843497e-06
```

**Comment by orbeckst** [comment]
@utkbansal I pushed a fix to your branch (! – this works on GitHub, see [Committing changes to a pull request branch created from a fork](https://help.github.com/articles/committing-changes-to-a-pull-request-branch-created-from-a-fork/)). 

The problem was twofold:
- `p_last` needs to refer to the last frame in the trajectory and there needs to be a `universe.trajectory[-1]` to move it to the last frame
- The underlying universes for `p_first` and `p_last` *must be independent* but if we just use `u` as a fixture then this is the same for each function and therefore `p_first` and `p_last` will always refer to the same frame.

I added a test `test_p_frames` that is not strictly necessary for MDAnalysis but makes sure that the tests are not broken in the future by someone who wants to optimize away "redundant" code.

The `test_rms.py` now pass locally for me. Hopefully also on Travis.

**Comment by orbeckst** [comment]
I fixed a few other things that caught my eye. Please feel free to squash all my commits into yours.

**Pull request merged**